xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -228,86 +238,259 @@ private static LdsUpdate processClientSideListener(Listener listener, boolean pa
+  static FilterChain parseFilterChain(
+      io.envoyproxy.envoy.config.listener.v3.FilterChain proto, Set<String> rdsResources,
+      TlsContextManager tlsContextManager, FilterRegistry filterRegistry, boolean parseHttpFilters)
+      throws ResourceInvalidException {
+    io.grpc.xds.HttpConnectionManager httpConnectionManager = null;
+    HashSet<String> uniqueNames = new HashSet<>();
+    for (io.envoyproxy.envoy.config.listener.v3.Filter filter : proto.getFiltersList()) {
+      if (!uniqueNames.add(filter.getName())) {
+        throw new ResourceInvalidException(
+            "FilterChain " + proto.getName() + " with duplicated filter: " + filter.getName());
+      }
+      if (!filter.hasTypedConfig()) {
+        throw new ResourceInvalidException(
+            "FilterChain " + proto.getName() + " contains filter " + filter.getName()
+                + " without typed_config");
+      }
+      Any any = filter.getTypedConfig();
+      // HttpConnectionManager is the only supported network filter at the moment.
+      if (!any.getTypeUrl().equals(TYPE_URL_HTTP_CONNECTION_MANAGER)) {
+        throw new ResourceInvalidException(
+            "FilterChain " + proto.getName() + " contains filter " + filter.getName()
+                + " with unsupported typed_config type " + any.getTypeUrl());
+      }
+      // Any filters after the first HttpConnectionManager are ignored.
+      if (httpConnectionManager == null) {
+        HttpConnectionManager hcmProto;
+        try {
+          hcmProto = any.unpack(HttpConnectionManager.class);
+        } catch (InvalidProtocolBufferException e) {
+          throw new ResourceInvalidException("FilterChain " + proto.getName() + " with filter "
+              + filter.getName() + " failed to unpack message", e);
+        }
+        httpConnectionManager = parseHttpConnectionManager(
+            hcmProto, rdsResources, filterRegistry, parseHttpFilters, false /* isForClient */);
+      }
+    }
+    if (httpConnectionManager == null) {
+      throw new ResourceInvalidException("FilterChain " + proto.getName()
+          + " missing required HttpConnectionManager filter");
+    }
+
+    EnvoyServerProtoData.DownstreamTlsContext downstreamTlsContext = null;
+    if (TRANSPORT_SOCKET_NAME_TLS.equals(proto.getTransportSocket().getName())) {
+      DownstreamTlsContext downstreamTlsContextProto;
+      try {
+        downstreamTlsContextProto =
+            proto.getTransportSocket().getTypedConfig().unpack(DownstreamTlsContext.class);
+      } catch (InvalidProtocolBufferException e) {
+        throw new ResourceInvalidException("FilterChain " + proto.getName()
+            + " failed to unpack message", e);
+      }
+      downstreamTlsContext =
+          EnvoyServerProtoData.DownstreamTlsContext.fromEnvoyProtoDownstreamTlsContext(
+              downstreamTlsContextProto);
+    }
 
+    String name = proto.getName();
+    if (name.isEmpty()) {
+      name = UUID.randomUUID().toString();
```

Comment:
The cross language design may have suggested using UUID. But the listener-name followed by the index of the filterChain as the allocated name might be better for debuggability.

Reply:
That sounds to be true. But, let's focus on the correctness first and then improve/polish the debuggability. That should be a fairly easy local change, with code spanning only two private methods.